code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/ContentCardUIProvider.kt

```diff
@@ -0,0 +1,160 @@
+  OF ANY KIND, either express or implied. See the License for the specific language
+  governing permissions and limitations under the License.
+*/
+
+package com.adobe.marketing.mobile.messaging
+
+import com.adobe.marketing.mobile.AdobeCallbackWithError
+import com.adobe.marketing.mobile.AdobeError
+import com.adobe.marketing.mobile.Messaging
+import com.adobe.marketing.mobile.aepcomposeui.AepUI
+import com.adobe.marketing.mobile.aepcomposeui.contentprovider.AepUIContentProvider
+import com.adobe.marketing.mobile.aepcomposeui.uimodels.AepUITemplate
+import com.adobe.marketing.mobile.messaging.ContentCardSchemaDataUtils.buildTemplate
+import com.adobe.marketing.mobile.messaging.ContentCardSchemaDataUtils.getAepUI
+import com.adobe.marketing.mobile.services.Log
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.launch
+
+/**
+ * ContentCardUiProvider is responsible for fetching and managing the content for a given surface.
+ * It uses Adobe Messaging APIs to retrieve propositions and transform them into UI templates for display.
+ *
+ * @property surfaceString The identifier of the surface for which content needs to be fetched.
+ */
+class ContentCardUIProvider(val surfaceString: String) : AepUIContentProvider {
+    companion object {
+        private const val SELF_TAG: String = "ContentCardUIProvider"
+    }
+
+    private val _contentFlow = MutableStateFlow<List<AepUITemplate>>(emptyList())
+    private val contentFlow: StateFlow<List<AepUITemplate>> = _contentFlow
+
+    private val _aepUiFlow = MutableStateFlow<List<AepUI<*, *>>>(emptyList())
+    private val aepUiFlow: StateFlow<List<AepUI<*, *>>> = _aepUiFlow
+
+    /**
+     * Retrieves a flow of AepUI instances for the given surface.
+     *
+     * This function initiates the content fetch using [getContent] and then returns a flow of
+     * [AepUI] instances that represent the UI templates. The flow emits updates whenever new
+     * content is fetched or any changes occur.
+     *
+     * @return A [Flow] that emits a list of [AepUI] instances.
+     */
+    suspend fun getAepUi(): Flow<List<AepUI<*, *>>> {
+        getContent()
+        return aepUiFlow
+    }
+
+    /**
+     * Clears the current content and re-fetches new content for the given surface.
+     */
+    override suspend fun refreshContent() {
+        _contentFlow.value = emptyList()
+        _aepUiFlow.value = emptyList()
+        getContent()
+    }
+
+    /**
+     * Retrieves a flow of AepUITemplate lists for the given surface.
+     * The flow emits updates whenever new content is fetched.
+     *
+     * @return A flow that emits lists of AepUITemplate.
+     */
+    override suspend fun getContent(): Flow<List<AepUITemplate>> {
+        CoroutineScope(Dispatchers.IO).launch {
```

Comment:
Do you think we can reuse a single scope stored at class level? Would that be more efficient?
@prudrabhat what do you think?

Reply:
Given that this is a suspending function, and the caller will already have to invoke and launch this in a CoroutineScope, we can write the business logic directly. The caller can decide what scope to run this on.